### ISSUE TITLE
Closures inside structs can be flagged as redundant when they shouldn't be.

### ISSUE THREAD

**Issue by nebulaeandstars** [rust-lang/rust-clippy#7672]
using clippy 0.1.52 (07e0e2e 2021-03-24)

I tried this code:

```rust
pub struct Foo {
    command: Box<dyn Fn() -> String>,
}

impl Foo {
    pub fn new() -> Self {
        Self {
            command: Box::new(|| String::new()),
        }
    }
}
```

I expected the code to pass all targets.

Instead, this happened:

`cargo clippy` recommends that I replace `|| String::new()` with `String::new()`, which isn't correct and won't compile.

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
binary: rustc
commit-hash: 07e0e2ec268c140e607e1ac7f49f145612d0f597
commit-date: 2021-03-24
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

**Comment by giraffate** [comment]
Clippy suggests `String::new`, not `String::new()`, in the `redundant_closure` lint. And it can compile.

I tried it on the [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7079366deee941d4fe512fd5453b4942).

**Comment by nebulaeandstars** [comment]
You're right, I must have misread the hint.
